src/lib/Serialize.hs

```diff
@@ -104,10 +117,81 @@ prettyVal val = case val of
     return $ align $ group innerDoc
   atom -> return $ prettyPrec atom LowestPrec
 
+-- === taking memory snapshots for serializing heap-backed dex values ===
+
+data WithSnapshot a = WithSnapshot a [PtrSnapshot]  deriving Generic
+type RawPtr = Ptr ()
+-- TODO: we could consider using some mmap-able instead of ByteString
+data PtrSnapshot = ByteArray BS.ByteString
+                 | PtrArray [PtrSnapshot]
+                 | NullPtr  deriving Generic
+
+class HasPtrs a where
+  traversePtrs :: Applicative f => (PtrType -> RawPtr -> f RawPtr) -> a -> f a
+
+takeSnapshot :: HasPtrs a => a -> IO (WithSnapshot a)
+takeSnapshot x =
+  liftM (WithSnapshot x) $ execWriterT $ flip traversePtrs x \ptrTy ptrVal -> do
+    snapshot <- lift $ takePtrSnapshot ptrTy ptrVal
+    tell [snapshot]
+    return ptrVal
+
+restoreSnapshot :: HasPtrs a => WithSnapshot a -> IO a
+restoreSnapshot (WithSnapshot x snapshots) =
+  flip evalStateT snapshots $ flip traversePtrs x \_ _ -> do
+    (s:ss) <- get
+    put ss
+    lift $ restorePtrSnapshot s
+
+takePtrSnapshot :: PtrType -> RawPtr -> IO PtrSnapshot
+takePtrSnapshot _ ptrVal | ptrVal == nullPtr = return NullPtr
+takePtrSnapshot (_, ptrTy) ptrVal = case ptrTy of
+  PtrType eltTy -> do
+    childPtrs <- loadPtrPtrs ptrVal
+    PtrArray <$> mapM (takePtrSnapshot eltTy) childPtrs
+  _ -> ByteArray <$> loadPtrBytes ptrVal
```

Comment:
This requires some more care, or it will segfault when given a CUDA pointer. We should think extra hard about caching in presence of multiple backends, which might require different pointers to end up on different devices. This change might make it effectively impossible to use the CUDA backend, because it will probably fail to load the prelude.

Reply:
Good point. To handle CUDA pointers we'll need a way to track the sizes of GPU allocations. You've suggested before that the CUDA allocator might already provide an API for that. For now I'll at least add a check that it's a host pointer and throw "not implemented" if not.